client/libra-dev/src/error.rs

```diff
@@ -0,0 +1,31 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use once_cell::sync::Lazy;
+use std::borrow::{Borrow, BorrowMut};
+use std::cmp::min;
+use std::os::raw::c_char;
+
+const MAX_ERROR_LENGTH: usize = 1024;
+
+static mut LAST_ERROR: Lazy<[u8; MAX_ERROR_LENGTH]> = Lazy::new(|| [0u8; MAX_ERROR_LENGTH]);
+
+/// Update the most recent error.
+pub fn update_last_error(err: String) {
+    let last_error: &mut [u8; MAX_ERROR_LENGTH] = unsafe { LAST_ERROR.borrow_mut() };
+    let slice_str = err.into_bytes().into_boxed_slice();
+    let min_error_length = min(slice_str.len(), MAX_ERROR_LENGTH - 1);
+    last_error[..min_error_length].copy_from_slice(&slice_str[..min_error_length]);
+    // null terminate the string
+    last_error[min_error_length] = 0;
+}
+
+pub fn clear_error() {
+    let last_error: &mut [u8; 1024] = unsafe { LAST_ERROR.borrow_mut() };
+    last_error[0] = 0;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn libra_strerror() -> *const c_char {
+    LAST_ERROR.borrow().as_ptr().cast()
+}
```

Comment:
It looks like this error handling isn't sound and ends up being undefined behavior due to it being possible for multiple threads to simultaneously have mutable access to the `LAST_ERROR` variable. If we want to keep this `get_last_error` mechanism then theres a few things that will need to be changed in order to make this sound. The following comments are based on the assumption that we want to keep this same API.

* Define `LAST_ERROR` as a `static` and not as a `static mut`. It's generally an anti-pattern to use `static mut` since it gives the whole program mut access to the variable (although you have to access it via unsafe) which makes it much harder to ensure your program is correct.
* For something like this, you don't really need `Lazy` since initializing an empty buffer is pretty trivial, so drop the dep on `once_cell`.
* The goal is to have a 1024 byte buffer that we can stick a string in. We're going to need to modify the buffer so we need to have mut access at some point, but we have a `static` (not a `static mut`) so we only have a shared ref to the buffer. This means we'll need to use a type that can do dynamic borrow checking and we want the error message to probably be preserved on a per-thread basis:
```lang=rust
const MAX_ERROR_LENGTH: usize = 1024;
thread_local! {
    static LAST_ERROR: RefCell<[u8; MAX_ERROR_LENGTH]> = RefCell::new([0u8; MAX_ERROR_LENGTH]);
}
```

If we want to keep this "last error" mechanism I'd take a closer look at [this implementation](https://michael-f-bryan.github.io/rust-ffi-guide/errors/return_types.html) which goes the extra mile to make sure the API doesn't have to worry about the caller holding onto the pointer to the last error between calls into the library. (Its probably bad form to give a caller a `*const char` and then mutate that out from under them later, which is what this API currently does)

Reply:
As mentioned above the thread_local storage doesn’t quite work since we have no control over threading on c calls. 

Maybe we should simply declare the c api is not thread safe and ask user to deal with it instead.

Reply:
I'm not sure what you mean. C's normal `errno` is also thread local, so this behavior can just be documented. If you want the error info, you have to call `strerror()` on the same thread as the call.

Do the current implementations not work with this thread local setup?

Reply:
I thought we had trouble with thread_local before , that's why we endup choosing global singleton.

I don't have an strong idea on why we didn't make thread_local work though, maybe we can try again if you strongly favors it

Reply:
Yes, please try it again. The caller of this code would have to be doing something very strange to be grabbing errors from a different thread than made the call.

Reply:
I prefer we make that change separate ,  I propose:

1. we add some comment to data.h say for now C-API is not thread-safe yet
2. I will file an task and work on thread_local! implementation to make it thread-safe soon.

It is a bigger change that don't really affect current use cases, but delaying checking into the tree however is causing some issues on new feature development.

Reply:
IMO The code left as-is cannot be merged due to the API being unsound. We have an extremely high bar for having code which uses unsafe to be merged into the project. Though I'll let @metajack give the final verdict.

i I'm also really beginning to question the use of a single global buffer due to potential UB. If a reference to this single global buffer is passed into C land, then you probably need to ensure that isn't not being used anymore before you go to mutate that buffer again. The document that's been linked a few times really should be consulted when building an error mechanism like this as there are a number of subtle things that are difficult to implement correctly. If you note they take great care to ensure that the C side passes in a buffer where the message can be copied to in order to avoid worrying about using unsafe and potential UB when it comes to having both shared and mutable references to an object.

Reply:
The global buffer is only used for storing an error message, which is being accessed as an const char* to the user. existing library like libc has similiar design (http://man7.org/linux/man-pages/man3/strerror.3.html) notice that it even just gives back an char * with no const modifier.  There are _l _r variant later but this pattern is not something that I invented on the fly. However it looks like the actual implement of that function  is keeping dangling pointer instead of an global buffer.  

Yes I am aware the cavet on the issue of this function, but it is not used anywhere else beside FFI library itself, so I don't see what the harm this would do to existing codebase. However if you do feel strongly about it, maybe we should simply remove this error mechanism first and checkin the rest of it.